src/encoding.rs

```diff
@@ -2,6 +2,502 @@
+    }
+}
+
+pub enum ExemplarEncoder<'a> {
+    Text(text::ExemplarEncoder<'a>),
+    Protobuf(proto::ExemplarEncoder<'a>),
+}
+
+impl<'a> From<text::ExemplarEncoder<'a>> for ExemplarEncoder<'a> {
+    fn from(e: text::ExemplarEncoder<'a>) -> Self {
+        Self::Text(e)
+    }
+}
+impl<'a> From<proto::ExemplarEncoder<'a>> for ExemplarEncoder<'a> {
+    fn from(e: proto::ExemplarEncoder<'a>) -> Self {
+        Self::Protobuf(e)
+    }
+}
+
+impl<'a> ExemplarEncoder<'a> {
+    pub fn encode_exemplar<S: EncodeLabelSet, V: EncodeExemplarValue>(
+        &mut self,
+        exemplar: &Exemplar<S, V>,
+    ) -> Result<(), std::fmt::Error> {
+        match self {
+            Self::Text(e) => e.encode_exemplar(exemplar),
+            Self::Protobuf(e) => e.encode_exemplar(exemplar),
+        }
+    }
+
+    pub fn no_exemplar(&mut self) -> Result<(), std::fmt::Error> {
+        match self {
+            Self::Text(e) => e.no_exemplar(),
+            Self::Protobuf(e) => e.no_exemplar(),
+        }
+    }
+}
+
+pub trait EncodeLabelSet {
+    /// Encode the given instance in the OpenMetrics text encoding.
+    fn encode(&self, encoder: LabelSetEncoder) -> Result<(), std::fmt::Error>;
+}
+
+impl<'a> From<text::LabelSetEncoder<'a>> for LabelSetEncoder<'a> {
+    fn from(e: text::LabelSetEncoder<'a>) -> Self {
+        Self::Text(e)
+    }
+}
+
+pub enum LabelSetEncoder<'a> {
+    Text(text::LabelSetEncoder<'a>),
+    Protobuf(proto::LabelSetEncoder<'a>),
+}
+
+impl<'a> From<proto::LabelSetEncoder<'a>> for LabelSetEncoder<'a> {
+    fn from(e: proto::LabelSetEncoder<'a>) -> Self {
+        Self::Protobuf(e)
+    }
+}
+
+impl<'a> LabelSetEncoder<'a> {
+    pub fn encode_label(&mut self) -> LabelEncoder {
+        match self {
+            LabelSetEncoder::Text(e) => e.encode_label().into(),
+            LabelSetEncoder::Protobuf(e) => e.encode_label().into(),
+        }
+    }
+}
+
+pub trait EncodeLabel {
+    fn encode(&self, encoder: LabelEncoder) -> Result<(), std::fmt::Error>;
+}
+
+pub enum LabelEncoder<'a> {
+    Text(text::LabelEncoder<'a>),
+    Protobuf(proto::LabelEncoder<'a>),
+}
+
+impl<'a> From<text::LabelEncoder<'a>> for LabelEncoder<'a> {
+    fn from(e: text::LabelEncoder<'a>) -> Self {
+        Self::Text(e)
+    }
+}
+
+impl<'a> From<proto::LabelEncoder<'a>> for LabelEncoder<'a> {
+    fn from(e: proto::LabelEncoder<'a>) -> Self {
+        Self::Protobuf(e)
+    }
+}
+
+impl<'a> LabelEncoder<'a> {
+    pub fn encode_label_key(&mut self) -> LabelKeyEncoder {
+        match self {
+            LabelEncoder::Text(e) => e.encode_label_key().into(),
+            LabelEncoder::Protobuf(e) => e.encode_label_key().into(),
+        }
+    }
+}
+
+pub trait EncodeLabelKey {
+    fn encode(&self, encoder: &mut LabelKeyEncoder) -> Result<(), std::fmt::Error>;
+}
+
+pub enum LabelKeyEncoder<'a> {
+    Text(text::LabelKeyEncoder<'a>),
+    Protobuf(proto::LabelKeyEncoder<'a>),
+}
+
+impl<'a> From<text::LabelKeyEncoder<'a>> for LabelKeyEncoder<'a> {
```

Comment:
I'd expect that these are only used within this crate (and not exported publicly) so I am not sure we need these impls.